src/bootstrap/format.rs

```diff
@@ -101,19 +107,58 @@ pub fn format(build: &Build, check: bool) {
     }
     let ignore_fmt = ignore_fmt.build().unwrap();
 
-    let rustfmt_path = build.config.initial_rustfmt.as_ref().unwrap_or_else(|| {
-        eprintln!("./x.py fmt is not supported on this channel");
-        std::process::exit(1);
+    let rustfmt_path = build
+        .config
+        .initial_rustfmt
+        .as_ref()
+        .unwrap_or_else(|| {
+            eprintln!("./x.py fmt is not supported on this channel");
+            std::process::exit(1);
+        })
+        .to_path_buf();
+    let src = build.src.clone();
+    let (tx, rx): (SyncSender<PathBuf>, _) = std::sync::mpsc::sync_channel(128);
+    let walker =
+        WalkBuilder::new(src.clone()).types(matcher).overrides(ignore_fmt).build_parallel();
+
+    // there is a lot of blocking involved in spawning a child process and reading files to format.
+    // spawn more processes than available cores to keep the CPU busy
+    let max_processes = num_cpus::get() * 2;
+
+    // spawn child processes on a separate thread so we can batch entries we have received from ignore
+    let thread = std::thread::spawn(move || {
+        let mut children = VecDeque::new();
+        while let Ok(path) = rx.recv() {
+            // try getting a few more paths from the channel to amortize the overhead of spawning processes
+            let paths: Vec<_> = rx.try_iter().take(7).chain(std::iter::once(path)).collect();
+
+            let child = rustfmt(&src, &rustfmt_path, paths.as_slice(), check);
+            children.push_back(child);
+
+            if children.len() > max_processes {
+                // await oldest child
+                children.pop_front().unwrap()();
+            }
+        }
+
+        // await remaining children
+        for mut child in children {
+            child();
+        }
     });
-    let src = &build.src;
-    let walker = WalkBuilder::new(src).types(matcher).overrides(ignore_fmt).build_parallel();
+
     walker.run(|| {
+        let tx = tx.clone();
         Box::new(move |entry| {
             let entry = t!(entry);
             if entry.file_type().map_or(false, |t| t.is_file()) {
-                rustfmt(src, &rustfmt_path, &entry.path(), check);
+                t!(tx.send(entry.into_path()));
             }
             ignore::WalkState::Continue
         })
     });
+
+    drop(tx);
```

Comment:
Why do you explicitly drop `tx` here?

Reply:
to close the channel, otherwise the thread would wait for new messages forever and not terminate.

Reply:
Does that mean if an error occurs `tidy` will hang forever, because tx gets dropped last?

Reply:
Not sure what scenario you have in mind. But if walkdir panics then the program will terminate anyway since it's on the main thread. If the `rustfmt`-spawning thread panics then `rx` gets dropped which should result in `tx` encountering an error and terminating the walk.

Reply:
@jyn514 It is a pattern when using channels such since if it does not explicitly dropped (or wait for it to goes out of scope), I see this pattern quite often. I wanted to comment if we should put this as a comment to show those does never seen this but I didn't? Maybe we should drop a comment saying that we need to drop the last transfer reference count so that the receiving part will stop waiting?

By the way, happy Chinese New Year!

Reply:
I think such an explanation/example would be more appropriate somewhere in the `std::sync` module if it's not there already.

Reply:
Looks like @pickfire put up PR #80269 to serve as that requested explanation.